### PR TITLE
feat: persist remote sessions + lazy reconnect across app restart

### LINKED ISSUE #12: Remote SSH [4/8]: Persistence + lazy reconnect across app restart
## Parent PRD

#8

## What to build

Make remote sessions survive an app restart without opening any network connections up front. At startup, remote sessions materialize in a `pending` state showing a cached preview of their last known output. On the first user click, cc-pewpew opens the host's control connection, probes the remote tmux session, and either reattaches or marks the session dead. Once a host is connected, all other sessions on that host are probed eagerly.

See the Reconnection and UI sections of PRD #8 for the design.

## Acceptance criteria

- [ ] `sessions.json` records persist `hostId` and a compact `lastKnownState` field (most recent thumbnail digest/text + timestamp)
- [ ] At app startup, remote sessions are materialized in `pending` state without opening any SSH connections
- [ ] Session cards render with the cached preview and a grey status dot while `pending`
- [ ] First click on a pending remote session opens the host's control connection, runs `tmux has-session`, and either reattaches the PTY (session transitions to `live`) or marks the session dead
- [ ] Once a host is connected, all other persisted sessions on that host are probed eagerly (single batched ssh acceptable)
- [ ] New states surfaced in the UI: `pending`, `host-unreachable`, `host-auth-failed`, each with distinct visual treatment
- [ ] A manual "Reconnect" action is available on sessions in non-live states; no automatic retries in v1
- [ ] SSH authentication failures are distinguished from network failures at reconnect time via the ssh-exit-signature parser; auth failures transition directly to `host-auth-failed` with no further attempts
- [ ] No background ssh traffic while the app is idle (verified by observation — no control connections to hosts with only `pending` sessions)
- [ ] Local session restore behavior unchanged

## Blocked by

- Blocked by #11

## User stories addressed

- User story 20
- User story 21
- User story 22
- User story 23
- User story 24
- User story 27
- User story 28
- User story 48
- User story 52

### PR DESCRIPTION
**Pull request by pmatos**
## Summary

- Remote sessions materialize in a new `pending` connection state at restart with a cached pane-text preview, without opening any SSH connections
- First click on a pending card opens the host's control connection, runs `tmux has-session`, and either reattaches the PTY (→ `live`) or marks the session dead
- After first connect, remaining pending siblings on the same host are probed in a single batch over the live ControlMaster — no new SSH handshakes
- New states `host-auth-failed` (red) and `host-unreachable` (amber) with distinct CSS + DetailPane overlays; manual Reconnect on any non-live remote session; no auto-retry in v1
- `sessions.json` gains `lastKnownState` (≤3 KiB, rate-limited to 10s) so the cached preview survives restarts
- Local session restore behavior unchanged

Closes #12

## Design notes

- `reconnectRemoteSession(id)` is probe-only — reattaches on tmux-present, marks dead on tmux-absent. `reviveSession` keeps its create-if-missing semantics for the truly-dead case.
- Auth vs. network is read from `runtimeStateFor(hostId)` (exported from `host-connection`), which is set by `startRuntime` before `ensureHostConnection` rejects — no re-parsing stderr.
- Idempotency: module-level in-flight Maps for both `reconnectRemoteSession(id)` (per session) and `probePendingSessionsOnHost(hostId)` (per host). Concurrent clicks coalesce.
- Auth-failed cascade: `probePendingSessionsOnHost` short-circuits when the runtime is already `auth-failed` or `unreachable`, marking all siblings without any network I/O.
- Orphaned `hostId` (host deleted from registry): `reconnectRemoteSession` sets `connectionState='unreachable'` and skips SSH.
- `lastKnownState` is captured from the local `PtyEntry.buffer` we already accumulate — no new remote I/O. Remote thumbnail capture remains deferred to issue #13.

## Test plan

- [x] `npx tsc --noEmit` — clean
- [x] `npx eslint .` — clean
- [x] `npx prettier --check .` — clean
- [x] `npx vitest run` — 177 pass (19 new in `session-manager.test.ts` covering: pending materialization, preserved prior status, running→idle normalization, dead preserved, lastKnownState round-trip, local-restore regression, reconnectRemoteSession (tmux-present, tmux-gone, auth-failed, unreachable, orphaned host, idempotency, retry-after-auth-failed), batch probe (parallel, tmux-gone-sibling, auth-failed cascade, idempotency), lastKnownState rate-limit + 3 KiB cap)
- [x] `npm run build` — clean
- [ ] Manual CDP walkthrough per PRD #8 reconnection flow (host unreachable, reconnect button, auth-failed retry)

### CODE REVIEW

File: src/main/session-manager.ts [hunk 1]
```diff
@@ -582,6 +602,164 @@
...
+    } else {
+      session.connectionState = 'offline'
+    }
+    onSessionsChanged()
+    if (retained) {
+      await releaseHostConnection(hostId).catch(() => undefined)
+    }
+    throw err
+  }
+  if (retained) {
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main/session-manager.ts]: Fixed in `d730a8c`

File: src/main/session-manager.ts [hunk 2]
```diff
@@ -582,6 +602,164 @@ export async function killSession(id: string): Promise<void> {
...
+    try {
+      const alive = await hasRemoteTmuxSession(s.id, host)
+      if (alive) {
+        await reattachRemotePty(s.id, host)
+        s.connectionState = 'live'
+        if (s.status === 'running') s.status = 'idle'
+        s.lastActivity = Date.now()
+      } else {
+        s.connectionState = 'offline'
+        s.status = 'dead'
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main/session-manager.ts]: Fixed in `d730a8c`. Added `probeRemoteTmuxSession` returning `'present' | 'absent' | 'unreachable'`. Both the single-session reconnect and the batch probe now mark `unreachable` on SSH-level failures (timeout / network / auth) instead of downgrading to `dead`. The batch additionally bails on an unreachable probe so the remaining siblings stay `pending` for a later manual retry.

File: src/main/session-manager.ts [hunk 3]
```diff
@@ -582,6 +602,164 @@ export async function killSession(id: string): Promise<void> {
...
+    await promise
+  } finally {
+    inflightReconnects.delete(id)
+  }
+  // Fire-and-forget the sibling batch probe — the caller should not block on
+  // it. `probePendingSessionsOnHost` is idempotent so concurrent clicks on
+  // multiple cards of the same host still collapse to a single batch.
+  const entry = sessions.get(id)
+  if (entry?.session.hostId && entry.session.connectionState === 'live') {
+    probePendingSessionsOnHost(entry.session.hostId).catch((err) => {
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main/session-manager.ts]: Fixed in `d730a8c`. The batch probe now fires whenever `runtimeStateFor(hostId) === 'live'` after a reconnect, regardless of whether the clicked session itself reached `live`. If the first click lands on a session whose remote tmux has been killed, the host's control connection is still established, so siblings on that host are still reconciled.

File: src/main/session-manager.ts [hunk 4]
```diff
@@ -582,6 +603,184 @@ export async function killSession(id: string): Promise<void> {
...
+  if (existing) return existing
+
+  const promise = doReconnectRemoteSession(id)
+  inflightReconnects.set(id, promise)
+  try {
+    await promise
+  } finally {
+    inflightReconnects.delete(id)
+  }
+  // Fire-and-forget the sibling batch probe — the caller should not block on
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main/session-manager.ts]: Fixed in `1eb9010`. `reconnectRemoteSession` now captures the rejection from `doReconnectRemoteSession`, fires the batch probe if the host's runtime state is known (success or failure), then re-throws the original error. The batch's existing short-circuit cascades `auth-failed` / `unreachable` to every pending sibling on the host without any new SSH — matches spec AC #8. Added a dedicated test asserting auth-failed cascade with exactly one `ensureHostConnection` call and zero `exec` calls.

File: src/main/session-manager.ts [hunk 5]
```diff
@@ -582,6 +603,192 @@ export async function killSession(id: string): Promise<void> {
...
+      onSessionsChanged()
+    }
+  } catch (err) {
+    const runtimeState = runtimeStateFor(hostId)
+    if (runtimeState === 'auth-failed') {
+      session.connectionState = 'auth-failed'
+    } else if (runtimeState === 'unreachable') {
+      session.connectionState = 'unreachable'
+    } else {
+      session.connectionState = 'offline'
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main/session-manager.ts:796]: Great catch — genuine production bug. Fixed in `39a1733`.

`prepareRemoteHost` now captures `runtimeStateFor(host.hostId)` *before* calling `stopHostConnection` (which deletes the runtime entry) and attaches it to the thrown error as `hostConnectionState`. `doReconnectRemoteSession` reads the tagged state first and falls back to `runtimeStateFor` only when the failure happened after `prepareRemoteHost` returned (e.g. bootstrap / attach). `probePendingSessionsOnHost` gained an optional `stateHint` plumbed through from the reconnect site, so the auth-failed / unreachable cascade short-circuit works even when the runtime entry is gone.

The reason our existing auth-failed tests didn't catch this: the `stopHostConnection` mock was a no-op, so production's delete-on-teardown was invisible to tests. Updated the mock to actually delete from the shared state map — that turned the existing auth-failed tests red and forced the fix.

File: src/main/session-manager.ts [hunk 6]
```diff
@@ -582,6 +615,216 @@ export async function killSession(id: string): Promise<void> {
...
+  if (hostId) {
+    // Prefer the state tagged onto the error by prepareRemoteHost — the
+    // runtime may have been deleted by stopHostConnection, so a direct
+    // runtimeStateFor lookup here would be `undefined` for the very cases
+    // (auth-failed / unreachable) the batch's short-circuit needs to catch.
+    const tagged = (reconnectError as { hostConnectionState?: HostConnectionState } | null)
+      ?.hostConnectionState
+    const stateHint = tagged ?? runtimeStateFor(hostId)
+    if (stateHint) {
+      probePendingSessionsOnHost(hostId, stateHint).catch((err) => {
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main/session-manager.ts]: Another sharp catch — fixed in `ffb5083`. On the `'absent'` and `'unreachable'` paths the PTY never retains the runtime, so the final `releaseHostConnection` inside `doReconnectRemoteSession` drops refcount to zero → `stopHostConnection` → runtime entry deleted before the outer caller ever reads `runtimeStateFor`. `doReconnectRemoteSession` now snapshots the state BEFORE the release and returns it; `reconnectRemoteSession` threads it into `probePendingSessionsOnHost` as `stateHint`. Also updated the test mocks to actually model refcount (retain bumps, release drops, zero → delete) — the previous mock was a no-op pair which is exactly why the bug was invisible to tests. New regression test asserts the sibling still reaches `live` after `state.runtimeStates.has('h1') === false`.

File: src/main/index.ts [hunk 7]
```diff
@@ -533,12 +545,20 @@ app.whenReady().then(async () => {
...
+  // any new SSH traffic (issue #12 AC #1 / #9). updateLastKnownState rate-limits
+  // internally so the 3s tick doesn't churn sessions.json.
   const thumbInterval = setInterval(() => {
     const thumbs = captureThumbnails()
     if (Object.keys(thumbs).length > 0) {
       broadcastToAll('thumbnails:text-updated', thumbs)
     }
+    for (const session of getSessions()) {
+      const text = getLastSnapshot(session.id)
+      if (text) updateLastKnownState(session.id, text)
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main/index.ts]: Fixed in `ffb5083`. Added `updateLastKnownStatesBatch(updates)` — applies the 10s per-session rate-limit to every entry in memory, then fires a single `onSessionsChanged()` (= one `sessions.json` write + one `sessions:updated` broadcast) when *any* update was accepted. The 3s thumbnail interval in `index.ts` now collects all `getLastSnapshot`-backed updates into an array and calls the batch once. Two tests assert (a) one broadcast for an N-session batch, and (b) zero broadcasts when all updates are rate-limited.

File: src/main/session-manager.ts [hunk 8]
```diff
@@ -582,6 +642,236 @@ export async function killSession(id: string): Promise<void> {
...
+  // leaves it false throws inside prepareRemoteHost and goes through the
+  // catch block's `throw err`.
+  //
+  // Snapshot the runtime state BEFORE the release: on `absent` or
+  // `unreachable` outcomes the PTY did not take a retain, so this release
+  // drops the refcount to zero and stopHostConnection deletes the entry. The
+  // outer reconnectRemoteSession uses the returned state as the sibling
+  // batch probe's stateHint.
+  const finalState = runtimeStateFor(hostId)
+  await releaseHostConnection(hostId).catch(() => undefined)
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main/session-manager.ts]: Fixed in `5a3cead`. `doReconnectRemoteSession` now returns `{state, retainedForBatch: true}` on the happy path and defers the release. `reconnectRemoteSession` wraps `probePendingSessionsOnHost` in try/finally inside the fire-and-forget closure; on finally the retain is released exactly once. This keeps the ControlMaster alive across the boundary so sibling probes reuse the existing `ControlPath` — one `ssh` handshake total, not N. On the `present` path the sibling's own reattach retain means the post-batch release drops from refs=2 to refs=1, leaving the runtime alive for the live PTYs. Two new tests cover: (a) absent + live sibling (refs=1 post-batch, exactly one `ensureHostConnection` call), (b) absent + no sibling (runtime torn down at refs=0).

File: src/main/session-manager.ts [hunk 9]
```diff
@@ -582,6 +642,254 @@ export async function killSession(id: string): Promise<void> {
...
+    for (const s of pending) s.connectionState = runtime
+    onSessionsChanged()
+    return
+  }
+
+  let bailed = false
+  for (const s of pending) {
+    if (bailed) break
+    try {
+      const probe = await probeRemoteTmuxSession(s.id, host)
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main/session-manager.ts:878]: Fixed in `00abd28`. `doProbePendingSessionsOnHost` now rechecks `s.connectionState !== 'pending'` at the top of each iteration and skips — so a concurrent reconnect that advanced a sibling out of pending mid-batch doesn't cause a duplicate `reattachRemotePty` / leaked refcount. New test wires a side-effect callback on the first probe that flips the second sibling to `live` (simulating the concurrent click) and asserts the batch skips its reattach while still processing the first and third.

File: src/main/session-manager.ts [hunk 10]
```diff
@@ -582,6 +642,259 @@ export async function killSession(id: string): Promise<void> {
...
+  // - on success (runtime was `live`), we reconcile siblings over the
+  //   now-live ControlMaster
+  // - on auth-failed / unreachable, the batch's short-circuit cascades that
+  //   state to every pending sibling without any new SSH I/O (spec AC #8)
+  //
+  // Skip only when there's no host at all (orphaned hostId / missing registry
+  // entry) or we couldn't determine any state — there's nothing to probe.
+  const entry = sessions.get(id)
+  const hostId = entry?.session.hostId
+  const tagged = (reconnectError as { hostConnectionState?: HostConnectionState } | null)
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main/session-manager.ts]: Fixed in `6fca3b6`. `reconnectRemoteSession` now captures `initialHostId` from `sessions.get(id)` *before* the await on `doReconnectRemoteSession` and uses it as a fallback after the await: `hostId = sessions.get(id)?.session.hostId ?? initialHostId`. If a concurrent `removeSession(id)` deletes the map entry while reconnect is in flight, the post-await fallback still gives us the host so the batch/release path fires and the ControlMaster retain is released.

New regression test (`releases host retain when session is removed mid-reconnect`) gates `ensureHostConnection`, calls `removeSession('r1')` while the reconnect is parked, then asserts `runtimeRefs.has('h1') === false` after the fire-and-forget release completes. Without the fallback the test fails with refs=1 (verified locally).

File: src/renderer/components/DetailPane.tsx [hunk 11]
```diff
@@ -15,7 +15,15 @@ export default function DetailPane({ sessionId, sessionName, onClose }: Props) {
   const hosts = useHostsStore((s) => s.hosts)
   const host = session?.hostId ? hosts.find((h) => h.hostId === session.hostId) : null
   const isDead = session?.status === 'dead'
+  const connectionState = session?.connectionState
+  const isRemote = !!session?.hostId
+  const isPending = isRemote && connectionState === 'pending'
+  const isConnecting = isRemote && connectionState === 'connecting'
+  const isAuthFailed = isRemote && connectionState === 'auth-failed'
+  const isUnreachable = isRemote && connectionState === 'unreachable'
+  const showReconnectOverlay = !isDead && (isPending || isAuthFailed || isUnreachable)
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/renderer/components/DetailPane.tsx]: Fixed in `6a7bbf3`. `DetailPane` now adds `isOffline = isRemote && connectionState === 'offline'` to the `showReconnectOverlay` set with a dedicated message ("Reconnect failed — The session could not be reattached on {host}. Click Retry to try again."). The same Retry button reuses `handleReconnect`, so users get the same recovery path as `auth-failed` / `unreachable` instead of an inert Terminal pane.

File: src/main/session-manager.ts [hunk 12]
```diff
@@ -237,6 +251,52 @@ function updateSession(id: string, status: SessionStatus): void {
...
+// to collapse the disk write + broadcast into one call (avoids an O(N) write
+// storm from a tight timer loop).
+function applyLastKnownState(id: string, text: string, now: number): boolean {
+  const entry = sessions.get(id)
+  if (!entry) return false
+  const last = lastKnownStateWrites.get(id) ?? 0
+  if (now - last < LAST_KNOWN_STATE_MIN_INTERVAL_MS) return false
+  const trimmed =
+    text.length > LAST_KNOWN_STATE_MAX_BYTES ? text.slice(-LAST_KNOWN_STATE_MAX_BYTES) : text
+  entry.session.lastKnownState = { text: trimmed, timestamp: now }
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main/session-manager.ts:277]: Fixed in `6a7bbf3`. `applyLastKnownState` now returns `false` when the trimmed text equals the existing `lastKnownState.text`, before mutating the entry or recording the rate-limit timestamp. Idle sessions that emit identical thumbnail text every 3s tick now no-op past the 10s window instead of churning `sessions.json` indefinitely. Not bumping `lastKnownStateWrites` on the no-op also lets a real text change after a long stable period write immediately rather than waiting another 10s.

New regression test (`skips write when text is unchanged across rate-limit windows`) advances time past the rate-limit, asserts zero broadcasts on identical text, and confirms a subsequent text change does broadcast. Test fails (1 broadcast) without the equality early-return, verified locally.
